proposals/Proposal: add cookies.removeAll() method.md

```diff
@@ -0,0 +1,147 @@
+
+### Known Consumers
+
+All extensions that access and/or modify cookies with awareness of partitioned cookies, through the use of the `partitionKey`  property  in  the  `cookies` extension API.
+
+## Specification
+
+
+### Schema
+
+##### Syntax
+
+```
+let removed = await browser.cookies.removeAll(
+details // object
+)
+```
+##### Parameters
+An `object` containing information to identify the cookie(s) to remove. It contains the following properties:
+
+>[`name`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/remove#name) optional:
+>
+>A  `string`  representing the name of the cookie to remove.
+>
+>[`partitionKey`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/remove#partitionkey)  optional:
+>
+>An object representing the storage partition containing the cookie.
+>
+>[`topLevelSite`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/remove#toplevelsite)  optional:
+>
+>A  `string`  representing the first-party URL of the top-level site storage partition containing the cookie.
+>
+>[`storeId`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/remove#storeid)  optional:
+>
+>A  `string`  representing the ID of the cookie store to find the cookie in. If unspecified, the cookie is looked for by default in the current execution context's cookie store.
+>
+>[`url`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/remove#url) optional:
+>
+>A  `string`  representing the URL associated with the cookie.
+
+### Behavior
+
+The API will remove all cookies that match the `details` object parameter with the exception of the cases outlined in the implementation details.
+If the extension does not have  [host permissions](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions)  for this URL, the API call will fail.
+
+### New Permissions
+No  new  permissions  are  required.
+
+### Manifest File Changes
+No new manifest fields are required.
+
+## Security and Privacy
+Incorrect usage can lead to the removal of cookies that are not intended.
+
+### Exposed Sensitive Data
+None
+
+### Abuse Mitigations
+To remove a cookie, the extension must have host permissions for the `url` associated with the cookie.
+
+
+## Alternatives
+None
+
+### Existing Workarounds
+
+The behavior introduced by this API can be replicated by using a combination of `cookies.getAll()` and `cookies.remove()`.
+
+### Open Web API
+
+The cookies API is specific to extensions
+
+## Implementation Notes
+
+Host permissions are required for this API.
+#### Special cases:
+```cookies.removeAll({})```
+
+>An empty details object will result in all cookies being removed. 
```

Comment:
I'd like some input on this. My thought was to keep the behavior the same between `cookies.getAll` and this implementation but it might make sense to block it or direct developers to use the already existing browserData API

Reply:
I would also like you make `{}` be an error. In Safari, we only return cookies you have host permissions for in `getAll({})`, so that is what would be removed if we supported it with `removeAll({})`. However, it does seem to be too much of a foot gun.

I think requiring `url` is still useful and safe.

Reply:
Thank you for the feedback @xeenon and @oliverdunk , having `{}` be an error works for me.

I'm inclined to agree with @xeenon, that having at least one key is useful and safe. Would requiring the caller to provide at least a `url` or a `topLevelSite` sound good?

A `url` allows for a similar requirement as the existing `cookies.remove()` and it's inclusion allows for preserving the functionality. While a `topLevelSite` enables the new functionality of removing all cookies associated with a single `topLevelSite` introduced by the API. Keeping both of these parameters optional but requiring at least one of them, matches existing APIs such as [webNavigation.getFrame()](https://developer.chrome.com/docs/extensions/reference/api/webNavigation#method-getFrame) that already have the pattern of having all optional parameters but returning an error when no parameters or an invalid combination of parameters are provided.

Reply:
I'm comfortable with that. Thanks!